File: lib/phrase.py
#!/usr/bin/env python2.4
# -*- coding: utf-8 -*-
#  Copyright (C) 2007 Jacek Śliwerski (rzyjontko)
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; version 2.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software Foundation,
#  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  

import re
from stem import PorterStemmer, SnowballStemmer

stemmers = {
    'da' : SnowballStemmer('danish'),
    'de' : SnowballStemmer('german'),
    'en' : PorterStemmer(),
    'es' : SnowballStemmer('spanish'),
    'fi' : SnowballStemmer('finnish'),
    'fr' : SnowballStemmer('french'),
    'hu' : SnowballStemmer('hungarian'),
    'it' : SnowballStemmer('italian'),
    'nl' : SnowballStemmer('dutch'),
    'no' : SnowballStemmer('norwegian'),
    'pt' : SnowballStemmer('portuguese'),
    'ro' : SnowballStemmer('romanian'),
    'ru' : SnowballStemmer('russian'),
    'sv' : SnowballStemmer('swedish'),
    }


class GenericHandler:
    def __init__(self, connectors):
        self._connectors = sorted(connectors + ["'"])

    def discard(self, word):
        return word.startswith('%') \
            or word.startswith('<') \
            or word in self._connectors
        

class AFHandler(GenericHandler):
    def __init__(self):
        GenericHandler.__init__(self, ["'n", "die"])

class DAHandler(GenericHandler):
    def __init__(self):
        GenericHandler.__init__(self, ["de", "dem", "den", "det", "en", "et"])

class DEHandler(GenericHandler):
    def __init__(self):
        GenericHandler.__init__(self, ["das", "dem", "den", "der", "deren", "des", "dessen", "die", "ein", "eine", "einem", "einen"])

class ENHandler(GenericHandler):
    def __init__(self):
        GenericHandler.__init__(self, ["a", "an", "the"])
    
class ESHandler(GenericHandler):
    def __init__(self):
        GenericHandler.__init__(self, ["el", "la", "las", "los", "una", "uno", "unas", "unos"])

class FRHandler(GenericHandler):
    def __init__(self):
        GenericHandler.__init__(self, ["la", "le", "les", "un", "une"])

class ITHandler(GenericHandler):
    def __init__(self):
        GenericHandler.__init__(self, ["i", "il", "lo", "gli", "la", "le", "un", "uno", "una"])

class NLHandler(GenericHandler):
    def __init__(self):
        GenericHandler.__init__(self, ["de", "een", "het"])

class PLHandler(GenericHandler):
    def __init__(self):
        GenericHandler.__init__(self, ["by"])

class PTHandler(GenericHandler):
    def __init__(self):
        GenericHandler.__init__(self, ["o", "os", "a", "as", "um", "uns", "uma", "umas"])

class SVHandler(GenericHandler):
    def __init__(self):
        GenericHandler.__init__(self, ['de', 'den', 'det', 'en', 'ena', 'ett'])


class Phrase:
    wre = re.compile("([\w'%<]/?(?:[&'_]?\w)*>?)", re.UNICODE)
    dre = re.compile('^\d+$', re.UNICODE)
    xre = re.compile("<\?w*>", re.UNICODE)
    qre = re.compile('"[^"]*"', re.UNICODE)

    __handlers = { "C"  : ENHandler (),
                   "af" : AFHandler (),
                   "da" : DAHandler (),
                   "de" : DEHandler (),
                   "en" : ENHandler (),
                   "es" : ESHandler (),
                   "fr" : FRHandler (),
                   "it" : ITHandler (),
                   "nl" : NLHandler (),
                   "pl" : PLHandler (),
                   "pt" : PTHandler (),
                   "sv" : SVHandler ()}
    __def_handler = GenericHandler ([])

    def __resolve(self, lang):
        lang = lang[:2]
        if lang in self.__handlers:
            return self.__handlers[lang]
        else:
            return self.__def_handler

    def __filterfun(self, word, handler):
        return len(word) < 50 \
            and not re.match(self.dre, word) \
            and not re.match(self.xre, word) \
            and not handler.discard(word)

    def __init__(self, phrase, lang, sort=True, stem=False):
        handler = self.__resolve (lang)
        self._phrase = phrase
        self._wordlist = filter(lambda x: self.__filterfun(x, handler), \
                                map(lambda x: x.replace('_', '').lower(), \
                                    self.wre.findall(phrase)))[:10]
        if stem and lang[:2] in stemmers:
            stemmer = stemmers[lang[:2]]
            self._wordlist = [stemmer.stem(word) for word in self._wordlist]
        if sort:
            self._wordlist.sort()

    def length(self):
        return len(self._wordlist)
    
    def canonical(self):
        return reduce (lambda x, y: x + ' ' + y, self._wordlist, '').lstrip()

    def canonical_list(self):
        return self._wordlist

    def required(self):
        return [m[1:-1] for m in self.qre.findall(self._phrase)][:5]

File: lib/phrase.c
/* 
   Copyright (C) 2006 Jacek Sliwerski

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; version 2.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  

   ----------------------------------------------------------------------

*/
/****************************************************************************
 *    IMPLEMENTATION HEADERS
 ****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <wchar.h>
#include <wctype.h>

#include "xmalloc.h"
#include "phrase.h"

/****************************************************************************
 *    IMPLEMENTATION PRIVATE DEFINITIONS / ENUMERATIONS / SIMPLE TYPEDEFS
 ****************************************************************************/

#define MIN(a, b) (((a) < (b)) ? (a) : (b))

/****************************************************************************
 *    IMPLEMENTATION PRIVATE CLASS PROTOTYPES / EXTERNAL CLASS REFERENCES
 ****************************************************************************/
/****************************************************************************
 *    IMPLEMENTATION PRIVATE STRUCTURES / UTILITY CLASSES
 ****************************************************************************/

typedef struct list
{
        struct list *next;
        word_t *word;
} list_t;


/****************************************************************************
 *    IMPLEMENTATION REQUIRED EXTERNAL REFERENCES (AVOID)
 ****************************************************************************/
/****************************************************************************
 *    IMPLEMENTATION PRIVATE DATA
 ****************************************************************************/
/****************************************************************************
 *    INTERFACE DATA
 ****************************************************************************/
/****************************************************************************
 *    IMPLEMENTATION PRIVATE FUNCTION PROTOTYPES
 ****************************************************************************/

static void word_destroy (word_t *word);

/****************************************************************************
 *    IMPLEMENTATION PRIVATE FUNCTIONS
 ****************************************************************************/

static wchar_t *
utf8_to_wchar(const unsigned char *utf, size_t *memsize)
{
        int i = 0;
        int j = 0;
	int len = strlen ((char *) utf);
	wchar_t *mem = xcalloc (len + 1, sizeof(wchar_t));

	while (i < len)
	{
		if(utf[i] < 0x80)
			mem[j++] = utf[i++];
		else if(utf[i] < 0xE0)
		{
			mem[j++] = ((utf[i] & 0x1F) << 6) |
				(utf[i + 1] & 0x3F);
			i += 2;
		}
		else if(utf[i] < 0xF0)
		{
			mem[j++] = ((utf[i] & 0x0F) << 12) |
				((utf[i + 1] & 0x3F) << 6) |
				(utf[i + 2] & 0x3F);
			i += 3;
		}
		else if(utf[i] < 0xF8)
		{
			mem[j++] = ((utf[i] & 0x07) << 18) |
				((utf[i + 1] & 0x3F) << 12) |
				((utf[i + 2] & 0x3F) << 6) |
				(utf[i + 2] & 0x3F);
			i += 4;
		}
		else if(utf[i] < 0xFC)
		{
			mem[j++] = ((utf[i] & 0x03) << 24) |
				((utf[i + 1] & 0x3F) << 18) |
				((utf[i + 2] & 0x3F) << 12) |
				((utf[i + 3] & 0x3F) << 6) |
				(utf[i + 4] & 0x3F);
			i += 5;
		}
		else if(utf[i] >= 0xFC)
		{
			mem[j++] = ((utf[i] & 0x01) << 30) |
				((utf[i + 1] & 0x3F) << 24) |
				((utf[i + 2] & 0x3F) << 18) |
				((utf[i + 3] & 0x3F) << 12) |
				((utf[i + 4] & 0x3F) << 6) |
				(utf[i + 5] & 0x3F);
			i += 6;
		}
	}

	mem = xrealloc (mem, sizeof(wchar_t) * (j + 1));
	*memsize = j;
        return mem;
}


static void
skip_nonalpha (wchar_t **text)
{
        while (**text && !iswalpha(**text))
                (*text)++;
}


static void
skip_alpha (wchar_t **text)
{
        while (**text && iswalpha(**text))
                (*text)++;
}


static int
mark_duplicates (word_t **array, int count)
{
        int i, j;
        int dups = 0;

        for (i = 0; i < count - 1; i++){
                for (j = i + 1; j < count; j++){
                        if (phrase_compare ((phrase_t *)array[i],
                                            (phrase_t *)array[j]) == 0){
                                array[i]->occurences++;
                                dups++;
                                word_destroy (array[j]);
                        } else {
                                break;
                        }
                }
                i = j - 1;
        }
        return dups;
}


static int
words_compare (const void *p1, const void *p2)
{
        const word_t **word1 = (const word_t **) p1;
        const word_t **word2 = (const word_t **) p2;

        return phrase_compare ((phrase_t *) *word1, (phrase_t *) *word2);
}


static word_t **
eliminate (list_t *list, int *count)
{
        int i, j;
        int dups;
        word_t **result;
        word_t **array = alloca (*count * sizeof (word_t *));

        for (i = 0; list; i++, list = list->next){
                array[i] = list->word;
        }

        qsort (array, *count, sizeof (word_t *), words_compare);
        dups = mark_duplicates (array, *count);
        *count -= dups;
        result = xmalloc (*count * sizeof (word_t *));

        for (i = 0, j = 0; j < *count + dups; i++){
                result[i] = array[j];
                j += array[j]->occurences;
        }

        return result;
}


static word_t *
word_create (wchar_t *text, int length)
{
        int i;
        word_t *result = xmalloc (sizeof (word_t));
        result->text = text;
        result->length = length;
        result->occurences = 1;

        for (i = 0; i < length; i++)
                result->text[i] = towlower (result->text[i]);

        return result;
}


static void
word_destroy (word_t *word)
{
        xfree (word);
}


/****************************************************************************
 *    INTERFACE FUNCTIONS
 ****************************************************************************/


unsigned
phrase_hash (phrase_t *phrase)
{
        int i;
        unsigned h = 0;
        unsigned g;

        for (i = 0; i < phrase->length; i++){
                h <<= 4;
                h  += phrase->text[i];
                if ((g = h & 0xf0000000)){
                        h ^= (g >> 24);
                        h ^= g;
                }
        }
        return h;
}


int
phrase_compare (phrase_t *phrase1, phrase_t *phrase2)
{
        int len1 = phrase1->length;
        int len2 = phrase2->length;

        int ret = wcsncasecmp (phrase1->text, phrase2->text, MIN (len1, len2));

        if (ret != 0)
                return -ret;

        return (len1 < len2) - (len1 > len2);
}


void
phrase_print (phrase_t *phrase)
{
        int i;

        for (i = 0; i < phrase->length; i++){
                printf ("%lc", phrase->text[i]);
        }
}


phrase_t *
phrase_create (const char *text)
{
        int count = 0;
        size_t length;
        wchar_t *ptr;
        wchar_t *word;
        list_t *list = NULL;
        phrase_t *result = xmalloc (sizeof (phrase_t));

        result->text = utf8_to_wchar ((const unsigned char *) text, & length);
        result->length = length;

        ptr = result->text;

        do {
                skip_nonalpha (& ptr);
                word = ptr;
                skip_alpha (& ptr);
                if (*word){
                        list_t *node = alloca (sizeof (list_t));
                        node->word = word_create (word, ptr - word);
                        node->next = list;
                        list = node;
                        count++;
                }
        } while (*ptr);

        result->words = eliminate (list, & count);
        result->word_count = count;
        
        return result;
}


void
phrase_destroy (phrase_t *phrase)
{
        int i;

        for (i = 0; i < phrase->word_count; i++)
                word_destroy (phrase->words[i]);

        xfree (phrase->text);
        if (phrase->words)
                xfree (phrase->words);
        xfree (phrase);
}

/****************************************************************************
 *
 *    END MODULE phrase.c
 *
 ****************************************************************************/
